File: marketdna-backend/app/models/indicator_edge.py
"""Pydantic models for Indicator Edge Lab."""
from typing import Optional
from pydantic import BaseModel


class SignalStats(BaseModel):
    occurrences: int
    avg_1w: float
    avg_2w: float
    avg_1m: float
    median_1m: float
    max_1m: float
    min_1m: float
    std_1m: float
    win_rate: float
    positive_prob: float
    negative_prob: float
    max_drawdown_1m: float       # avg max adverse excursion during holding period
    distribution: dict[str, int] # return buckets: "<-10", "-10:-5", "-5:0", "0:5", "5:10", ">10"
    expected_value: float
    # Stock regime-conditional stats — None when < MIN_SAMPLES in that bucket
    regime_high_win_rate: Optional[float] = None   # win rate when stock regime ≥ 60 at signal time
    regime_low_win_rate:  Optional[float] = None   # win rate when stock regime < 60 at signal time
    regime_high_ev:       Optional[float] = None   # EV when stock regime ≥ 60
    regime_low_ev:        Optional[float] = None   # EV when stock regime < 60
    regime_high_occurrences: int = 0
    regime_low_occurrences:  int = 0
    # Market regime-conditional stats (avg regime across all 50 NIFTY stocks)
    mkt_regime_high_win_rate: Optional[float] = None
    mkt_regime_low_win_rate:  Optional[float] = None
    mkt_regime_high_ev:       Optional[float] = None
    mkt_regime_low_ev:        Optional[float] = None
    mkt_regime_high_occurrences: int = 0
    mkt_regime_low_occurrences:  int = 0


class ThresholdResult(BaseModel):
    threshold: float
    label: str
    occurrences: int
    avg_1m: float
    win_rate: float
    expected_value: float


class IndicatorEdgeResult(BaseModel):
    name: str
    signal: str
    category: str
    direction: str          # "Bullish" | "Bearish" | "Mixed"
    stats: SignalStats
    grade: str              # A+, A, B, C, D
    edge_score: float
    threshold_sweep: list[ThresholdResult]
    best_threshold: ThresholdResult | None
    is_active: bool


class CurrentSignal(BaseModel):
    indicator: str
    current_value: str
    signal_label: str
    is_active: bool
    stats: SignalStats | None


class IndicatorEdgeReport(BaseModel):
    symbol: str
    data_start: str
    data_end: str
    total_bars: int
    results: list[IndicatorEdgeResult]
    ranked_indicators: list[str]
    best_indicator: str
    best_setup: str
    current_signals: list[CurrentSignal]
    computed_at: str


class StockEdgeSummaryItem(BaseModel):
    symbol: str
    best_indicator: str
    best_signal: str
    grade: str
    direction: str
    edge_score: float
    win_rate: float
    avg_1m: float
    negative_prob: float
    max_drawdown_1m: float
    expected_value: float
    occurrences: int
    is_active: bool


class StockEdgeSummaryResult(BaseModel):
    items: list[StockEdgeSummaryItem]
    total_symbols: int
    computed_at: str

File: marketdna-web/src/types/indicatorEdge.ts
export interface SignalStats {
  occurrences: number
  avg_1w: number
  avg_2w: number
  avg_1m: number
  median_1m: number
  max_1m: number
  min_1m: number
  std_1m: number
  win_rate: number
  positive_prob: number
  negative_prob: number
  max_drawdown_1m: number
  distribution: Record<string, number>
  expected_value: number
  // Stock regime-conditional stats (null = < 5 occurrences in that bucket)
  regime_high_win_rate: number | null
  regime_low_win_rate: number | null
  regime_high_ev: number | null
  regime_low_ev: number | null
  regime_high_occurrences: number
  regime_low_occurrences: number
  // Market regime-conditional stats (avg regime across all 50 NIFTY stocks)
  mkt_regime_high_win_rate: number | null
  mkt_regime_low_win_rate: number | null
  mkt_regime_high_ev: number | null
  mkt_regime_low_ev: number | null
  mkt_regime_high_occurrences: number
  mkt_regime_low_occurrences: number
}

export interface ThresholdResult {
  threshold: number
  label: string
  occurrences: number
  avg_1m: number
  win_rate: number
  expected_value: number
}

export interface IndicatorEdgeResult {
  name: string
  signal: string
  category: string
  direction: string
  stats: SignalStats
  grade: string
  edge_score: number
  threshold_sweep: ThresholdResult[]
  best_threshold: ThresholdResult | null
  is_active: boolean
}

export interface CurrentSignal {
  indicator: string
  current_value: string
  signal_label: string
  is_active: boolean
  stats: SignalStats | null
}

export interface IndicatorEdgeReport {
  symbol: string
  data_start: string
  data_end: string
  total_bars: number
  results: IndicatorEdgeResult[]
  ranked_indicators: string[]
  best_indicator: string
  best_setup: string
  current_signals: CurrentSignal[]
  computed_at: string
}

export interface StockEdgeSummaryItem {
  symbol: string
  best_indicator: string
  best_signal: string
  grade: string
  direction: string
  edge_score: number
  win_rate: number
  avg_1m: number
  negative_prob: number
  max_drawdown_1m: number
  expected_value: number
  occurrences: number
  is_active: boolean
}

export interface StockEdgeSummaryResult {
  items: StockEdgeSummaryItem[]
  total_symbols: number
  computed_at: string
}

export const GRADE_COLOR: Record<string, string> = {
  'A+': '#22c55e',
  'A':  '#86efac',
  'B':  '#fbbf24',
  'C':  '#94a3b8',
  'D':  '#ef4444',
}

export const DIR_COLOR: Record<string, string> = {
  Bullish: '#22c55e',
  Bearish: '#ef4444',
  Mixed:   '#fbbf24',
}
